torch/utils/data/datapipes/_typing.py

```diff
@@ -365,6 +365,23 @@ def conditional_reset(*args, **kwargs):
 
             namespace['reset'] = conditional_reset
 
+        if '__len__' in namespace:
+            len_func = namespace['__len__']
+        else:
+            len_func = None
+
+        def len_hook(*args, **kwargs):
+            datapipe = args[0]
+            if datapipe._length is not None:
+                return datapipe._length
+            else:
+                if len_func is not None:
+                    return len_func(*args, **kwargs)
+                else:
+                    raise TypeError(f"object of type '{datapipe}' has no len()")
```

Comment:
The original exception is a `TypeError` but I can switch it to `NotImplementedError` if that is preferred.

Reply:
`TypeError` is needed for running `list(dp)`.